File: custom_components/eink_calendar/renderer/generate_icons.js
#!/usr/bin/env node
/**
 * Generate PNG icons from Material Design Icons (MDI) SVG files.
 *
 * This script downloads SVG icons from the @mdi/svg npm package and converts
 * them to 24x24 PNG files for use in the E-Ink Calendar renderer.
 */

const fs = require('fs');
const path = require('path');
const https = require('https');
const { createCanvas, loadImage } = require('canvas');

// Icon names to generate (without mdi: prefix)
const ICONS_TO_GENERATE = [
  // Existing icons (for reference/regeneration)
  'airplane',
  'bell',
  'briefcase',
  'calendar',
  'car',
  'check',
  'email',
  'food',
  'gift',
  'heart',
  'home',
  'medical',
  'party',
  'phone',
  'school',
  'shopping',
  'star',
  // New common calendar icons
  'infinity',
  'account',
  'account-group',
  'baby-face',
  'bank',
  'basket',
  'beach',
  'beer',
  'bike',
  'book',
  'bookmark',
  'brush',
  'bug',
  'cake',
  'camera',
  'cash',
  'charity',
  'chart-line',
  'church',
  'city',
  'clipboard',
  'clock',
  'coffee',
  'cog',
  'compass',
  'controller-classic',
  'creation',
  'domain',
  'dumbbell',
  'earth',
  'family-tree',
  'ferry',
  'file-document',
  'film',
  'finance',
  'fire',
  'flag',
  'flask',
  'flower',
  'football',
  'forest',
  'fuel',
  'gamepad',
  'glass-cocktail',
  'golf',
  'hammer',
  'handshake',
  'hiking',
  'hospital',
  'image',
  'karate',
  'key',
  'ladder',
  'laptop',
  'leaf',
  'library',
  'lightbulb',
  'link',
  'lock',
  'map-marker',
  'meditation',
  'microphone',
  'monitor',
  'motorbike',
  'music',
  'nature',
  'newspaper',
  'notebook',
  'palette',
  'paw',
  'pencil',
  'pharmacy',
  'piano',
  'pill',
  'pizza',
  'presentation',
  'printer',
  'puzzle',
  'racquetball',
  'restore',
  'robot',
  'rocket',
  'run',
  'sail-boat',
  'scale-balance',
  'scissors',
  'shield',
  'ship-wheel',
  'sitemap',
  'soccer',
  'sofa',
  'spa',
  'subway',
  'swim',
  'sync',
  'taxi',
  'teach',
  'television',
  'tennis',
  'thumb-up',
  'tools',
  'train',
  'trophy',
  'truck',
  'umbrella',
  'video',
  'wallet',
  'water',
  'waves',
  'weight-lifter',
  'wifi',
  'wrench',
  'yoga',
];

// Icon size (24x24 to match existing icons)
const ICON_SIZE = 24;

/**
 * Download MDI SVG icon from jsdelivr CDN
 */
async function downloadSvg(iconName, svgDir) {
  const svgFile = path.join(svgDir, `${iconName}.svg`);

  // Skip if already exists
  if (fs.existsSync(svgFile)) {
    return svgFile;
  }

  const url = `https://cdn.jsdelivr.net/npm/@mdi/svg@latest/svg/${iconName}.svg`;

  return new Promise((resolve, reject) => {
    console.log(`Downloading ${iconName}...`);

    https.get(url, (res) => {
      if (res.statusCode !== 200) {
        console.log(`  Warning: Failed to download ${iconName} (HTTP ${res.statusCode})`);
        resolve(null);
        return;
      }

      const chunks = [];
      res.on('data', (chunk) => chunks.push(chunk));
      res.on('end', () => {
        const svgContent = Buffer.concat(chunks).toString('utf8');

        if (!svgContent.includes('<svg')) {
          console.log(`  Warning: Invalid SVG for ${iconName}`);
          resolve(null);
          return;
        }

        fs.writeFileSync(svgFile, svgContent);
        resolve(svgFile);
      });
    }).on('error', (err) => {
      console.log(`  Error downloading ${iconName}: ${err.message}`);
      resolve(null);
    });
  });
}

/**
 * Convert SVG to PNG using node-canvas
 */
async function svgToPng(svgFile, pngFile) {
  try {
    const svgContent = fs.readFileSync(svgFile, 'utf8');

    // Extract viewBox or use default 24x24
    const viewBoxMatch = svgContent.match(/viewBox="([^"]+)"/);
    let viewBox = '0 0 24 24';
    if (viewBoxMatch) {
      viewBox = viewBoxMatch[1];
    }

    const [, , vbWidth, vbHeight] = viewBox.split(' ').map(Number);

    // Create a data URL from the SVG
    const svgDataUrl = `data:image/svg+xml;base64,${Buffer.from(svgContent).toString('base64')}`;

    // Load SVG as image
    const img = await loadImage(svgDataUrl);

    // Create canvas
    const canvas = createCanvas(ICON_SIZE, ICON_SIZE);
    const ctx = canvas.getContext('2d');

    // Draw the SVG scaled to fit
    ctx.drawImage(img, 0, 0, ICON_SIZE, ICON_SIZE);

    // Save as PNG
    const buffer = canvas.toBuffer('image/png');
    fs.writeFileSync(pngFile, buffer);

    return true;
  } catch (error) {
    console.log(`  Error converting ${path.basename(svgFile)}: ${error.message}`);
    return false;
  }
}

/**
 * Main function
 */
async function main() {
  const iconsDir = path.join(__dirname, 'icons');
  const svgDir = path.join(__dirname, 'icons_svg_cache');

  // Create directories
  if (!fs.existsSync(iconsDir)) {
    fs.mkdirSync(iconsDir);
  }
  if (!fs.existsSync(svgDir)) {
    fs.mkdirSync(svgDir);
  }

  console.log(`Generating ${ICONS_TO_GENERATE.length} icons...`);
  console.log(`Icons directory: ${iconsDir}`);
  console.log(`SVG cache directory: ${svgDir}`);
  console.log('');

  let successCount = 0;
  const failedIcons = [];

  for (const iconName of ICONS_TO_GENERATE) {
    const pngFile = path.join(iconsDir, `${iconName}.png`);

    // Download SVG
    const svgFile = await downloadSvg(iconName, svgDir);
    if (!svgFile) {
      failedIcons.push(iconName);
      continue;
    }

    // Convert to PNG
    if (await svgToPng(svgFile, pngFile)) {
      console.log(`  ✓ Generated ${iconName}.png`);
      successCount++;
    } else {
      failedIcons.push(iconName);
    }
  }

  console.log('');
  console.log(`Successfully generated ${successCount}/${ICONS_TO_GENERATE.length} icons`);

  if (failedIcons.length > 0) {
    console.log(`\nFailed to generate ${failedIcons.length} icons:`);
    failedIcons.forEach((icon) => console.log(`  - ${icon}`));
  }

  // Generate a JSON file listing all available icons
  const availableIcons = fs.readdirSync(iconsDir)
    .filter((f) => f.endsWith('.png'))
    .map((f) => f.replace('.png', ''))
    .sort();

  const iconsListFile = path.join(iconsDir, 'available_icons.json');
  fs.writeFileSync(iconsListFile, JSON.stringify(availableIcons, null, 2));

  console.log(`\nTotal icons available: ${availableIcons.length}`);
  console.log(`Icon list saved to: ${iconsListFile}`);
}

main().catch(console.error);

File: custom_components/eink_calendar/renderer/generate_icons.py
#!/usr/bin/env python3
"""
Generate PNG icons from Material Design Icons (MDI) SVG files.

This script downloads SVG icons from the @mdi/svg npm package and converts
them to 24x24 PNG files for use in the E-Ink Calendar renderer.
"""

import json
import subprocess
import sys
from pathlib import Path

try:
    from cairosvg import svg2png  # pyright: ignore[reportMissingImports]
except ImportError:
    print("Error: cairosvg is required. Install with: pip install cairosvg")
    sys.exit(1)

# Icon names to generate (without mdi: prefix)
ICONS_TO_GENERATE = [
    # Existing icons (for reference/regeneration)
    "airplane",
    "bell",
    "briefcase",
    "calendar",
    "car",
    "check",
    "email",
    "food",
    "gift",
    "heart",
    "home",
    "medical",
    "party",
    "phone",
    "school",
    "shopping",
    "star",
    # New common calendar icons
    "infinity",
    "account",
    "account-group",
    "baby-face",
    "bank",
    "basket",
    "beach",
    "beer",
    "bike",
    "book",
    "bookmark",
    "brush",
    "bug",
    "cake",
    "camera",
    "cash",
    "charity",
    "chart-line",
    "church",
    "city",
    "clipboard",
    "clock",
    "coffee",
    "cog",
    "compass",
    "controller-classic",
    "creation",
    "domain",
    "dumbbell",
    "earth",
    "family-tree",
    "ferry",
    "file-document",
    "film",
    "finance",
    "fire",
    "flag",
    "flask",
    "flower",
    "football",
    "forest",
    "fuel",
    "gamepad",
    "glass-cocktail",
    "golf",
    "hammer",
    "handshake",
    "hiking",
    "hospital",
    "image",
    "karate",
    "key",
    "ladder",
    "laptop",
    "leaf",
    "library",
    "lightbulb",
    "link",
    "lock",
    "map-marker",
    "meditation",
    "microphone",
    "monitor",
    "motorbike",
    "music",
    "nature",
    "newspaper",
    "notebook",
    "palette",
    "paw",
    "pencil",
    "pharmacy",
    "piano",
    "pill",
    "pizza",
    "presentation",
    "printer",
    "puzzle",
    "racquetball",
    "restore",
    "robot",
    "rocket",
    "run",
    "sail-boat",
    "scale-balance",
    "scissors",
    "shield",
    "ship-wheel",
    "sitemap",
    "soccer",
    "sofa",
    "spa",
    "subway",
    "swim",
    "sync",
    "taxi",
    "teach",
    "television",
    "tennis",
    "thumb-up",
    "tools",
    "train",
    "trophy",
    "truck",
    "umbrella",
    "video",
    "wallet",
    "water",
    "waves",
    "weight-lifter",
    "wifi",
    "wrench",
    "yoga",
]

# Icon size (24x24 to match existing icons)
ICON_SIZE = 24


def download_mdi_svg(icon_name: str, svg_dir: Path) -> Path | None:
    """
    Download MDI SVG icon from jsdelivr CDN.

    Args:
        icon_name: Icon name without 'mdi:' prefix (e.g., 'calendar', 'infinity')
        svg_dir: Directory to save SVG files

    Returns:
        Path to downloaded SVG file, or None if download failed
    """
    svg_file = svg_dir / f"{icon_name}.svg"

    if svg_file.exists():
        return svg_file

    # Try to download from jsdelivr CDN
    url = f"https://cdn.jsdelivr.net/npm/@mdi/svg@latest/svg/{icon_name}.svg"

    print(f"Downloading {icon_name}...")
    try:
        subprocess.run(
            ["curl", "-sL", "-o", str(svg_file), url],
            check=True,
            capture_output=True,
            text=True,
        )

        # Check if file was downloaded and is valid SVG
        if svg_file.exists() and svg_file.stat().st_size > 0:
            with open(svg_file, "r") as f:
                content = f.read()
                if "<svg" in content:
                    return svg_file

        print(f"  Warning: Failed to download {icon_name}")
        if svg_file.exists():
            svg_file.unlink()
        return None

    except subprocess.CalledProcessError as e:
        print(f"  Error downloading {icon_name}: {e}")
        return None


def convert_svg_to_png(svg_file: Path, png_file: Path) -> bool:
    """
    Convert SVG to PNG using cairosvg.

    Args:
        svg_file: Path to SVG file
        png_file: Path to output PNG file

    Returns:
        True if conversion succeeded, False otherwise
    """
    try:
        svg2png(
            url=str(svg_file),
            write_to=str(png_file),
            output_width=ICON_SIZE,
            output_height=ICON_SIZE,
        )
        return True
    except Exception as e:
        print(f"  Error converting {svg_file.name}: {e}")
        return False


def main():
    """Generate PNG icons from MDI SVG files."""
    script_dir = Path(__file__).parent
    icons_dir = script_dir / "icons"
    svg_dir = script_dir / "icons_svg_cache"

    # Create directories
    icons_dir.mkdir(exist_ok=True)
    svg_dir.mkdir(exist_ok=True)

    print(f"Generating {len(ICONS_TO_GENERATE)} icons...")
    print(f"Icons directory: {icons_dir}")
    print(f"SVG cache directory: {svg_dir}")
    print()

    success_count = 0
    failed_icons = []

    for icon_name in ICONS_TO_GENERATE:
        png_file = icons_dir / f"{icon_name}.png"

        # Download SVG
        svg_file = download_mdi_svg(icon_name, svg_dir)
        if not svg_file:
            failed_icons.append(icon_name)
            continue

        # Convert to PNG
        if convert_svg_to_png(svg_file, png_file):
            print(f"  ✓ Generated {icon_name}.png")
            success_count += 1
        else:
            failed_icons.append(icon_name)

    print()
    print(f"Successfully generated {success_count}/{len(ICONS_TO_GENERATE)} icons")

    if failed_icons:
        print(f"\nFailed to generate {len(failed_icons)} icons:")
        for icon_name in failed_icons:
            print(f"  - {icon_name}")

    # Generate a JSON file listing all available icons
    available_icons = sorted([f.stem for f in icons_dir.glob("*.png")])
    icons_list_file = icons_dir / "available_icons.json"
    with open(icons_list_file, "w") as f:
        json.dump(available_icons, f, indent=2)

    print(f"\nTotal icons available: {len(available_icons)}")
    print(f"Icon list saved to: {icons_list_file}")


if __name__ == "__main__":
    main()
